pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/FieldIndexConfigs.java

```diff
@@ -0,0 +1,121 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.pinot.segment.spi.index;
+
+import com.fasterxml.jackson.annotation.JsonAnyGetter;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.databind.JsonNode;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import org.apache.pinot.spi.config.table.IndexConfig;
+
+
+/**
+ * FieldIndexConfigs are a map like structure that relates index types with their configuration, providing a type safe
+ * interface.
+ *
+ * This class can be serialized into a JSON object whose keys are the index type ids using Jackson, but cannot be
+ * serialized back. A custom Jackson deserializer could be provided if needed.
+ */
+public class FieldIndexConfigs {
+
+  public static final FieldIndexConfigs EMPTY = new FieldIndexConfigs(new HashMap<>());
+
+  private final Map<IndexType, IndexConfig> _configMap;
+
+  private FieldIndexConfigs(Map<IndexType, IndexConfig> configMap) {
+    _configMap = Collections.unmodifiableMap(configMap);
+  }
+
+  /**
+   * Returns the configuration associated with the given index type, which will be null if there is no configuration for
+   * that index type.
+   */
+  @JsonIgnore
+  public <C extends IndexConfig, I extends IndexType<C, ?, ?>> C getConfig(I indexType) {
+    IndexConfig config = _configMap.get(indexType);
+    if (config == null) {
+      return indexType.getDefaultConfig();
+    }
+    return (C) config;
+  }
+
+  /*
+  This is used by Jackson when this object is serialized. Each entry of the map will be directly contained in the
+  JSON object, with the key name as the key in the JSON object and the result of serializing the key value as the value
+  in the JSON object.
+   */
+  @JsonAnyGetter
```

Comment:
This can be used, for example, to print a message to the user using Jackson. For example, something like:
```
LOGGER.log("bla bla bla bla", JsonUtils.objectToJson(objectOfTypeFieldIndexConfigs))
```

It can also be used to directly serialize this object into a JSON, for example if we want to recreate a `FieldConfig` from one of this objects

Reply:
It does confuse me when a class is not meant to be serialized but annotated. For debugging purpose, we should override the `toString()` method.

Reply:
Then the toString method would be quite more complex. I will add it, but it will indirectly delegate to this method.